fix: make SSR-safe and fix ESLint warning for Beacon in UserDropdown

albin-baby-2002 opened the pull request:
## What does this PR do?

This PR updates the UserDropdown component to be safe for server-side rendering (SSR) and addresses an ESLint warning related to a conditional expression.

- Added a guard `if (typeof window === "undefined") return;` to ensure `window.Beacon` is only accessed in the browser, preventing SSR errors.
- Replaced the `Beacon && Beacon(...)` expression with a proper `if (Beacon) { ... }` statement to satisfy ESLint rules (`@typescript-eslint/no-unused-expressions`).
- Added proper `BeaconFunction` type definition to the Window interface instead of using `@ts-ignore`.

- Fixes #24183 
- Fixes CAL-6493

## Updates since last revision

- **Merged upstream/main**: The `UserDropdown.tsx` file was moved from `packages/features/shell/user-dropdown/` to `apps/web/modules/shell/user-dropdown/` as part of an upstream refactor. This PR now targets the new file location.
- **Added dependency array to Beacon useEffect**: Per reviewer feedback, added `[user?.username]` as the dependency array to prevent the effect from running on every render while still updating session data when the username changes.
- **Added automated tests**: Created `UserDropdown.test.tsx` with 8 tests covering:
  - Beacon session-data calls with correct payload
  - Handling of undefined Beacon (no errors)
  - Lazy-loaded Beacon detection via polling
  - Username change triggering useEffect re-run
  - Component rendering states (loading, user available, no user)
- **Cleaned up test file** (per keithwillcode and Cubic feedback):
  - Removed redundant comments above vi.mock calls
  - Removed misleading SSR safety test that provided false coverage (jsdom always has window defined)
  - Removed unused `@ts-expect-error` directives (Window interface augmentation makes them unnecessary)
- **Fixed Beacon lazy loading issue** (per Cubic feedback): Implemented polling mechanism that checks every second for Beacon availability. When Beacon loads lazily after component mount, the polling detects it and sends session data, then clears the interval.

**Link to Devin run:** https://app.devin.ai/sessions/58c887a6e4f34d5dad0eaa09a212d733
**Requested by:** unknown ()

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [N/A] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

- Are there environment variables that should be set? N/A
- What are the minimal test data to have? N/A
- What is expected (happy path) to have (input and output)? N/A
- Any other important info that could help to test that PR: Run `yarn test apps/web/modules/shell/user-dropdown/UserDropdown.test.tsx` to execute the automated tests.

## Human Review Checklist

- [ ] Verify the polling mechanism (1-second interval) is appropriate for detecting lazy-loaded Beacon
- [ ] Confirm the interval cleanup on unmount prevents memory leaks
- [ ] Confirm the changes are correctly applied to `apps/web/modules/shell/user-dropdown/UserDropdown.tsx`
- [ ] Review that tests adequately cover the Beacon functionality including lazy loading scenario

albin-baby-2002 committed:
fix(user-dropdown): make SSR-safe and fix ESLint warning for Beacon

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/calcom/cal.com?pullRequest=24275) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/calcom/cal.com?pullRequest=24275) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/calcom/cal.com?pullRequest=24275) it.</sub>

pallava-joshi reviewed (changes requested):
dropped suggestions

albin-baby-2002 committed:
Merge remote-tracking branch 'upstream/main' into fix/user-dropdown-beacon-ssr

albin-baby-2002 committed:
fix: added ssr guard in second useEffect and added type for beacon instead of ts-ignoring

pallava-joshi reviewed (dismissed):
there's one more suggestion to fix useEffect's dependency.

albin-baby-2002 commented:
> there's one more suggestion to fix useEffect's dependency.

@Pallava-Joshi The issue is that if the useEffect runs before Beacon fully loads, the session data might not be sent. To handle this without affecting performance, I propose using a setInterval that periodically checks if Beacon is available. Once Beacon is ready, the session data is sent, and the interval is immediately cleared to avoid unnecessary calls. If the dependency user?.username changes later, the useEffect will trigger again, restarting the interval to ensure the updated session data is sent. This approach guarantees that session data is always sent when Beacon is ready while avoiding repeated or redundant calls.

devin-ai-integration[bot] committed:
Merge remote-tracking branch 'upstream/main' into fix/user-dropdown-beacon-ssr

devin-ai-integration[bot] committed:
fix: add dependency array to Beacon useEffect

Address review feedback by adding [user?.username] dependency array
to the Beacon session-data useEffect, improving performance by
preventing unnecessary re-runs on every render.

Co-authored-by: albin-baby-2002 <albinbaby2002@gmail.com>
Co-Authored-By: unknown <>

devin-ai-integration[bot] committed:
test: add automated tests for UserDropdown Beacon functionality

Add comprehensive tests for:
- Beacon session-data calls with correct payload
- Handling of undefined Beacon (no errors)
- Username change triggering useEffect re-run
- SSR safety (window undefined check)
- Component rendering states (loading, user available, no user)

Co-authored-by: albin-baby-2002 <albinbaby2002@gmail.com>
Co-Authored-By: unknown <>

keithwillcode commented on apps/web/modules/shell/user-dropdown/UserDropdown.test.tsx:
```diff
@@ -0,0 +1,246 @@
+import React from "react";
+import { vi, describe, it, expect, beforeEach, afterEach } from "vitest";
+
+// Mock next-auth/react
```
DevinAI remove these useless comments

devin-ai-integration[bot] committed:
refactor: remove useless comments and misleading SSR test

- Remove redundant comments above vi.mock calls (per keithwillcode feedback)
- Remove SSR safety test that provided false coverage since jsdom always has window defined (per Cubic feedback)
- The 'should not throw error when Beacon is undefined' test already covers graceful handling

Co-Authored-By: unknown <>

devin-ai-integration[bot] committed:
fix: remove unused @ts-expect-error directives

The Window interface is augmented with optional Beacon and Support properties,
so TypeScript allows deleting them without errors.

Co-Authored-By: unknown <>

devin-ai-integration[bot] committed:
fix: handle lazy-loaded Beacon with polling

When Beacon loads lazily after component mount, the useEffect now polls
every second until Beacon becomes available, then sends session data
and clears the interval. This ensures session info is always sent even
when HelpScout loads asynchronously.

Added test to verify lazy loading behavior works correctly.

Co-Authored-By: unknown <>
